### PR TITLE
[ncp] add radio only mode NCP app

### PR DESCRIPTION
**Pull request by bukepo**
This PR adds the radio-only NCP app.

The following changes are made,

- Fixed a crash in raw-link mode caused by passing NULL as `aFrame` to `otPlatRadioReceiveDone()` when error occurred during processing received frame
- Enhanced to start ACK timer when the transmission actually starts instead of the time to call `otPlatRadioTransmit()`
- Adjusted NCP buffer size and message pool size for radio-only mode app.
- Enable Soft ACK Timeout and Soft retransmission in radio-only mode under POSIX platform.
- Add missing APIs for radio-only mode.
- Move `otGetVersionString()` into `instance_api.cpp`.
- Move NEST_STREAM_MFG into `ncp_base.cpp`.

### CODE REVIEW

File: src/core/openthread-core-default-config.h [hunk 1]
```diff
@@ -675,7 +679,7 @@
  *
  */
 #ifndef OPENTHREAD_CONFIG_NCP_UART_RX_BUFFER_SIZE
-#define OPENTHREAD_CONFIG_NCP_UART_RX_BUFFER_SIZE               1300
+#define OPENTHREAD_CONFIG_NCP_UART_RX_BUFFER_SIZE               512
```
Comment (abtink) [inline src/core/openthread-core-default-config.h]: This seems unrelated to this commit.... Any reason to reduce the uart/spi buffer sizes?

File: src/core/openthread-core-default-config.h [hunk 2]
```diff
@@ -685,7 +689,7 @@
  *
  */
 #ifndef OPENTHREAD_CONFIG_NCP_SPI_BUFFER_SIZE
-#define OPENTHREAD_CONFIG_NCP_SPI_BUFFER_SIZE                   1300
+#define OPENTHREAD_CONFIG_NCP_SPI_BUFFER_SIZE                   512
```
Comment (abtink) [inline src/core/openthread-core-default-config.h]: Same question as above for SPI buffer...

File: include/openthread/link.h [hunk 3]
```diff
@@ -306,6 +306,19 @@ OTAPI void OTCALL otLinkSetPollPeriod(otInstance *aInstance, uint32_t aPollPerio
...
+ * @param[in] aInstance      A pointer to an OpenThread instance.
+ * @param[in] aShortAddress  The IEEE 802.15.4 Short Address.
+ *
+ * @retval OT_ERROR_NONE             If successful.
+ * @retval OT_ERROR_INVALID_STATE    If the raw link-layer isn't enabled.
+ *
+ */
+OTAPI otError OTCALL otLinkSetShortAddress(otInstance *aInstance, uint16_t aShortAddress);
+
+
```
Comment (abtink) [inline include/openthread/link.h]: nit: extra new line

File: configure.ac [hunk 4]
```diff
@@ -444,18 +444,28 @@ AC_ARG_ENABLE(ncp-app,
             enable_ncp_app=${enableval}
             enable_ncp_app_mtd=${enableval}
             enable_ncp_app_ftd=${enableval}
+            enable_ncp_app_radio=${enableval}
```
Comment (abtink) [inline configure.ac]: Can we consider combining the new new `--ncp-app=radio` mode with `--enable-raw-link-api`? 
What is the difference between the two options?

File: src/core/Makefile.am [hunk 5]
```diff
@@ -31,13 +31,18 @@ include $(abs_top_nlbuild_autotools_dir)/automake/pre.am
...
     $(NULL)
 
 CPPFLAGS_COMMON                     = \
     -I$(top_srcdir)/include           \
     $(OPENTHREAD_TARGET_DEFINES)      \
     $(NULL)
 
+libopenthread_radio_a_CPPFLAGS      = \
+    $(CPPFLAGS_COMMON)                \
+    -DOPENTHREAD_RADIO=1              \
```
Comment (abtink) [inline src/core/Makefile.am]: I know you are following the `OPENTHREAD_MTD` and `OPENTRHEAD_FTD` naming here.. but `OPENTRHEAD_RADIO` as a config define seems too generic and not very descriptive... I don't have a better suggestion for what to name this (will think about it)...

File: src/core/openthread-core-default-config.h (hunk 1 above)
Comment (bukepo) [inline src/core/openthread-core-default-config.h]: Thanks, I missed some code.

File: configure.ac (hunk 4 above)
Comment (bukepo) [inline configure.ac]: With `--ncp-app=radio`, a new app only has radio layer will be built. With `--enable-raw-link-api`, *raw link api* will be enabled in `ftd` and `mtd` mode app. And `--ncp-app=radio` implicitly enables *raw link api*.

File: src/core/Makefile.am (hunk 5 above)
Comment (bukepo) [inline src/core/Makefile.am]: Thanks. We also chewed about the name, but failed to came up with a better one. `raw` is too raw and `mac` is too thick. If `radio` also is inappropriate, how about `link`? @jwhui @abtink

File: src/core/api/instance_api.cpp [hunk 6]
```diff
@@ -290,3 +304,42 @@ otError otSetDynamicLogLevel(otInstance *aInstance, otLogLevel aLogLevel)
...
+
+#ifdef PLATFORM_VERSION_ATTR_PREFIX
+    PLATFORM_VERSION_ATTR_PREFIX
+#else
+    static
+#endif
+    const char sVersion[] =
+        PACKAGE_NAME "/" PACKAGE_VERSION
+#ifdef  PLATFORM_INFO
+        "; " PLATFORM_INFO
```
Comment (jwhui) [inline src/core/api/instance_api.cpp]: This version string was updated in #2077.

File: configure.ac (hunk 4 above)
Comment (abtink) [inline configure.ac]: Is there any use-case for enabling "raw-link-api" in ftd/mtd mode app?

File: src/core/Makefile.am (hunk 5 above)
Comment (abtink) [inline src/core/Makefile.am]: It will be only part of NCP app, right?
Some suggestions:
- `OPENTHREAD_RAW_LINK_NCP`
- `OPENTHREAD_RAW_LINK`

File: configure.ac (hunk 4 above)
Comment (bukepo) [inline configure.ac]: I'm not sure if there is any use-case. I think enabling *raw link api* in *ftd*/*mtd* mode allows a single firmware providing different layer APIs.

File: src/core/api/instance_api.cpp (hunk 6 above)
Comment (bukepo) [inline src/core/api/instance_api.cpp]: updated.

File: src/core/Makefile.am (hunk 5 above)
Comment (bukepo) [inline src/core/Makefile.am]: Basically we have two categories about the API provided through UART: cli and NCP. Then about the thread capabilities : mtd and ftd. I think the mode this PR added is more about capability than API. Potentially cli can implement this mode too. So I incline to name it similar to mtd and ftd. If we have to choose between the two you give, I prefer the latter one. Thoughts? @jwhui

**Comment by hubertmis** [comment]
Is `mMemoryPool` required in radio-only NCP? Removing this field from `otInstance` would decrease memory footprint.

**Comment by bukepo** [comment]
@hubertmis Thanks for your suggestion, I've removed the message pool.

**Comment by LuDuda** [comment]
Great work, nice to see this topic is ongoing. May i ask, how do you test this solution, specifically, what do you have on host side?

**Comment by bukepo** [comment]
I'm also preparing another PR for the host side, which will include tests for this solution.

File: examples/platforms/da15000/platform.c [hunk 7]
```diff
@@ -143,6 +143,7 @@ bool PlatformPseudoResetWasRequested(void)
     return false;
 }
 
+#if OPENTHREAD_MTD || OPENTHREAD_FTD
```
Comment (jwhui) [inline examples/platforms/da15000/platform.c]: How about feature flag this on the commissioner vs. joiner roles?

File: examples/platforms/da15000/platform.c [hunk 8]
```diff
@@ -277,13 +278,16 @@ static void StateChangedCallback(uint32_t aFlags, void *aContext)
...
     }
 }
+#endif // OPENTHREAD_MTD || OPENTHREAD_FTD
 
 void PlatformProcessDrivers(otInstance *aInstance)
 {
     if (sInstance == NULL)
     {
         sInstance = aInstance;
+#if OPENTHREAD_MTD || OPENTHREAD_FTD
```
Comment (jwhui) [inline examples/platforms/da15000/platform.c]: How about feature flag this on the commissioner vs. joiner roles?

File: src/core/api/link_raw_api.cpp [hunk 9]
```diff
@@ -357,13 +194,12 @@ void LinkRaw::InvokeReceiveDone(otRadioFrame *aFrame, otError aError)
         if (aError == OT_ERROR_NONE)
         {
             otLogInfoPlat(&mInstance, "LinkRaw Invoke Receive Done (%d bytes)", aFrame->mLength);
+            mReceiveDoneCallback(&mInstance, aFrame, aError);
```
Comment (jwhui) [inline src/core/api/link_raw_api.cpp]: We have a bunch of stats/counters in the MAC to record error occurrences.  If we do not want to pass all the errors to host, we'll need to consider adding the counters into the radio.

File: examples/platforms/da15000/platform.c (hunk 7 above)
Comment (bukepo) [reply examples/platforms/da15000/platform.c]: I actually removed notifier from radio only mode, because there's no thread state at all.
Besides, I'm wondering if we should divide platform code into two layers, app layer and driver layer. app layer should only call openthread upper layer API and driver layer should not.

Comment (jwhui) [reply examples/platforms/da15000/platform.c]: The point of this specific comment is that this code is not needed even in the MTD/FTD configurations if DTLS is not being used.

All that said, this is example code, so I'm okay to leave it as you have it.  As you noted, the platform code could be better designed.  The platform code here is intended to be just an example.

File: src/core/api/link_raw_api.cpp (hunk 9 above)
Comment (bukepo) [reply src/core/api/link_raw_api.cpp]: I'm also considering merging LinkRaw with Mac to reduce duplicated code after this feature passes all tests, which should take some additional efforts. Shall we deal with the counters later or in this PR?

Comment (jwhui) [reply src/core/api/link_raw_api.cpp]: I'm fine either way, what do you prefer?

Comment (bukepo) [reply src/core/api/link_raw_api.cpp]: I would prefer do it later, so that this feature could be available earlier.
